File: src/main/java/net/atos/zac/admin/ReferenceTableAdminService.java
/*
 * SPDX-FileCopyrightText: 2022 Atos, 2024 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.zac.admin;

import static jakarta.transaction.Transactional.TxType.REQUIRED;
import static jakarta.transaction.Transactional.TxType.SUPPORTS;
import static net.atos.zac.util.ValidationUtil.valideerObject;

import java.util.List;
import java.util.stream.Collectors;

import jakarta.enterprise.context.ApplicationScoped;
import jakarta.inject.Inject;
import jakarta.persistence.EntityManager;
import jakarta.persistence.PersistenceContext;
import jakarta.persistence.criteria.CriteriaBuilder;
import jakarta.persistence.criteria.CriteriaQuery;
import jakarta.persistence.criteria.Root;
import jakarta.transaction.Transactional;

import net.atos.zac.admin.model.HumanTaskReferentieTabel;
import net.atos.zac.admin.model.ReferenceTable;
import net.atos.zac.shared.exception.FoutmeldingException;

@ApplicationScoped
@Transactional(SUPPORTS)
public class ReferenceTableAdminService {
    private static final String UNIQUE_CONSTRAINT = "Er bestaat al een referentietabel met de code \"%s\".";
    private static final String FOREIGN_KEY_CONSTRAINT = "Deze referentietabel wordt gebruikt (voor: %s) en kan niet verwijderd worden.";

    @PersistenceContext(unitName = "ZaakafhandelcomponentPU")
    private EntityManager entityManager;

    private ReferenceTableService referenceTableService;

    /**
     * Default no-arg constructor, required by Weld.
     */
    public ReferenceTableAdminService() {
    }

    @Inject
    public ReferenceTableAdminService(
            final ReferenceTableService referenceTableService
    ) {
        this.referenceTableService = referenceTableService;
    }

    @Transactional(REQUIRED)
    public ReferenceTable newReferenceTable() {
        final ReferenceTable nieuw = new ReferenceTable();
        nieuw.setCode(getUniqueCodeForReferenceTable(1, referenceTableService.listReferenceTables()));
        nieuw.setNaam("Nieuwe referentietabel");
        return nieuw;
    }

    @Transactional(REQUIRED)
    public ReferenceTable createReferenceTable(final ReferenceTable referenceTable) {
        return updateReferenceTable(referenceTable);
    }

    @Transactional(REQUIRED)
    public ReferenceTable updateReferenceTable(final ReferenceTable referenceTable) {
        valideerObject(referenceTable);
        referenceTableService.findReferenceTable(referenceTable.getCode())
                .ifPresent(existing -> {
                    if (!existing.getId().equals(referenceTable.getId())) {
                        throw new FoutmeldingException(String.format(UNIQUE_CONSTRAINT, referenceTable.getCode()));
                    }
                });
        return entityManager.merge(referenceTable);
    }

    @Transactional(REQUIRED)
    public void deleteReferenceTable(final long id) {
        final ReferenceTable tabel = entityManager.find(ReferenceTable.class, id);
        final CriteriaBuilder builder = entityManager.getCriteriaBuilder();
        final CriteriaQuery<HumanTaskReferentieTabel> query = builder.createQuery(HumanTaskReferentieTabel.class);
        final Root<HumanTaskReferentieTabel> root = query.from(HumanTaskReferentieTabel.class);
        query.select(root).where(builder.equal(root.get("tabel").get("id"), tabel.getId()));
        final List<HumanTaskReferentieTabel> resultList = entityManager.createQuery(query).getResultList();
        if (!resultList.isEmpty()) {
            throw new FoutmeldingException(String.format(FOREIGN_KEY_CONSTRAINT, resultList.stream()
                    .map(HumanTaskReferentieTabel::getVeld)
                    .distinct()
                    .collect(Collectors.joining(", "))
            ));
        }
        entityManager.remove(tabel);
    }

    private String getUniqueCodeForReferenceTable(final int i, final List<ReferenceTable> list) {
        final String code = "TABEL" + (1 < i ? i : "");
        if (list.stream()
                .anyMatch(referentieTabel -> code.equals(referentieTabel.getCode()))) {
            return getUniqueCodeForReferenceTable(i + 1, list);
        }
        return code;
    }
}

File: src/main/kotlin/net/atos/zac/admin/ReferenceTableAdminService.kt
/*
 * SPDX-FileCopyrightText: 2024 Lifely
 * SPDX-License-Identifier: EUPL-1.2+
 */
package net.atos.zac.admin

import jakarta.enterprise.context.ApplicationScoped
import jakarta.inject.Inject
import jakarta.persistence.EntityManager
import jakarta.transaction.Transactional
import net.atos.zac.admin.model.HumanTaskReferentieTabel
import net.atos.zac.admin.model.ReferenceTable
import net.atos.zac.app.exception.InputValidationFailedException
import nl.info.zac.exception.ErrorCode.ERROR_CODE_REFERENCE_TABLE_IS_IN_USE_BY_ZAAKAFHANDELPARAMETERS
import nl.info.zac.exception.ErrorCode.ERROR_CODE_REFERENCE_TABLE_WITH_SAME_CODE_ALREADY_EXISTS
import nl.info.zac.exception.ErrorCode.ERROR_CODE_SYSTEM_REFERENCE_TABLE_CANNOT_BE_DELETED
import nl.info.zac.util.AllOpen
import nl.info.zac.util.NoArgConstructor

@ApplicationScoped
@Transactional(Transactional.TxType.SUPPORTS)
@NoArgConstructor
@AllOpen
class ReferenceTableAdminService @Inject constructor(
    private val entityManager: EntityManager,
    private val referenceTableService: ReferenceTableService
) {
    @Transactional(Transactional.TxType.REQUIRED)
    fun createReferenceTable(referenceTable: ReferenceTable): ReferenceTable {
        referenceTableService.findReferenceTable(referenceTable.code)?.run {
            throw InputValidationFailedException(ERROR_CODE_REFERENCE_TABLE_WITH_SAME_CODE_ALREADY_EXISTS)
        }
        return entityManager.merge(referenceTable)
    }

    @Transactional(Transactional.TxType.REQUIRED)
    fun updateReferenceTable(referenceTable: ReferenceTable): ReferenceTable {
        // check if there is already a different reference table, i.e. one with a different id, but with the same code
        referenceTableService.findReferenceTable(referenceTable.code)?.run {
            if (this.id != referenceTable.id) {
                throw InputValidationFailedException(ERROR_CODE_REFERENCE_TABLE_WITH_SAME_CODE_ALREADY_EXISTS)
            }
        }
        return entityManager.merge(referenceTable)
    }

    @Transactional(Transactional.TxType.REQUIRED)
    @Suppress("NestedBlockDepth")
    fun deleteReferenceTable(id: Long) {
        val referenceTable = referenceTableService.readReferenceTable(id).also {
            if (it.isSystemReferenceTable) {
                throw InputValidationFailedException(ERROR_CODE_SYSTEM_REFERENCE_TABLE_CANNOT_BE_DELETED)
            }
        }
        entityManager.criteriaBuilder.let { criteriaBuilder ->
            criteriaBuilder.createQuery(HumanTaskReferentieTabel::class.java).let { query ->
                query.from(HumanTaskReferentieTabel::class.java).let {
                    query.select(it)
                        .where(criteriaBuilder.equal(it.get<Any>("tabel").get<Any>("id"), referenceTable.id))
                }
                entityManager.createQuery(query).resultList.run {
                    if (this.isNotEmpty()) {
                        throw InputValidationFailedException(
                            ERROR_CODE_REFERENCE_TABLE_IS_IN_USE_BY_ZAAKAFHANDELPARAMETERS
                        )
                    }
                }
            }
            entityManager.remove(referenceTable)
        }
    }
}
